feat: Add mjpeg [screen] streaming

jzaleski opened the pull request:
While working on a project spanning both _Android_ and _iOS_ automation my team noticed that that there was `mjpeg` streaming support for _iOS_ but not _Android_. We built this internally and wanted to give it back to the open-source community.

This change-set adds mjpeg [screen] steaming support for _Android_, with reasonable defaults, configurable via the environment.

---

(Special thanks to @lc-applause for building this out, he made my job, prepping this for PR, easy)

mykola-mokhnach commented:
@jzaleski Thank you for your commitment.
Unfortunately, we have already tried this approach before and it failed miserably because the actual performance of `uia.takeScreenshot();` is pretty low. It takes about 200ms to take a single screenshot, which gives us maximum 5 FPS. 

Consider checking the recently implemented streaming feature based on screenrecord tool: https://github.com/appium/appium/blob/master/docs/en/writing-running-appium/android/android-screen-streaming.md

The adb-based approach above is capable to achieve 60 FPS at full HD

There is also https://github.com/Genymobile/scrcpy, which does not have some limitations that screenrecord has, although it is based on several private APIs

jzaleski commented:
Based on our testing, w/ Android SDK 25+, we were able to hit a consistent 20fps+ which was good enough for our use-case.

In the latest update to this PR you will notice we've [conservatively] defaulted the target-FPS to 10 (consistent with the default for iOS via WDA), adjusted a few other defaults and fixed a bug w/ the `targetInterval` calculation (used to lock the frame-rate).

Given the level of API compatibility I can understand the hesitation to merge this in as is. I'm fine w/ keeping this on our fork, but would you perhaps consider merging if it were disabled by default, as it is quite self-contained, and only enabled through settings / configuration?

jzaleski commented:
retest this please

mykola-mokhnach commented:
Could you please do small performance testing of the current implementation on both an emulator and a real device (like single frame duration, max FPS and average CPU usage) and attach the results? logcat output with timestamps should be fine

Thanks

jzaleski commented:
Will do @mykola-mokhnach, stay tuned. Will likely be tomorrow AM.

jzaleski commented:
_Found the time to get it done, here are the results..._

---

### Pixel (Emulator):

#### Screenshot capture `averageDurationMs`

```bash
04-15 17:44:58.238  8031  8055 I System.out: averageDurationMs over 30 screenshot(s): 47.466667
04-15 17:45:01.276  8031  8055 I System.out: averageDurationMs over 30 screenshot(s): 44.833332
04-15 17:45:04.331  8031  8055 I System.out: averageDurationMs over 30 screenshot(s): 44.4
04-15 17:45:07.377  8031  8055 I System.out: averageDurationMs over 30 screenshot(s): 44.2
04-15 17:45:10.417  8031  8055 I System.out: averageDurationMs over 30 screenshot(s): 45.3
04-15 17:45:13.461  8031  8055 I System.out: averageDurationMs over 30 screenshot(s): 46.766666
04-15 17:45:16.511  8031  8055 I System.out: averageDurationMs over 30 screenshot(s): 47.933334
04-15 17:45:19.572  8031  8055 I System.out: averageDurationMs over 30 screenshot(s): 51.366665
04-15 17:45:22.592  8031  8055 I System.out: averageDurationMs over 30 screenshot(s): 52.6
04-15 17:45:25.642  8031  8055 I System.out: averageDurationMs over 30 screenshot(s): 48.366665
```

(_The above was gathered by instrumenting:_ `matchFramerate` _in:_ MjpegScreenshotStream.java)

#### Current `master` - CPU & Memory

```bash
 8350 u0_a147      20   0  12G 108M  73M S  0.0   5.4   0:00.82 io.appium.uiaut+
 8350 u0_a147      20   0  12G 108M  73M S  0.0   5.4   0:00.82 io.appium.uiaut+
 8350 u0_a147      20   0  12G 108M  73M S  0.0   5.4   0:00.82 io.appium.uiaut+
 8350 u0_a147      20   0  12G 108M  73M S  0.0   5.4   0:00.83 io.appium.uiaut+
 8350 u0_a147      20   0  12G 108M  73M S  0.0   5.4   0:00.83 io.appium.uiaut+
```
 
#### With PR #340 applied (0 client(s)) - CPU & Memory
 
```bash
 8696 u0_a147      20   0  12G 109M  73M S  0.0   5.4   0:00.77 io.appium.uiaut+
 8696 u0_a147      20   0  12G 109M  73M S  0.0   5.4   0:00.78 io.appium.uiaut+
 8696 u0_a147      20   0  12G 109M  73M S  3.8   5.4   0:00.78 io.appium.uiaut+
 8696 u0_a147      20   0  12G 109M  73M S  0.0   5.4   0:00.79 io.appium.uiaut+
 8696 u0_a147      20   0  12G 109M  73M S  0.0   5.4   0:00.79 io.appium.uiaut+
```

#### With PR #340 applied (1 client(s)) - CPU & Memory
 
```bash
 8696 u0_a147      20   0  12G 126M  75M S 18.5   6.3   0:00.87 io.appium.uiaut+
 8696 u0_a147      20   0  12G 118M  75M S 15.3   5.8   0:01.05 io.appium.uiaut+
 8696 u0_a147      20   0  12G 113M  75M S 11.5   5.6   0:01.24 io.appium.uiaut+
 8696 u0_a147      20   0  12G 112M  75M S  7.6   5.6   0:01.41 io.appium.uiaut+
 8696 u0_a147      20   0  12G 118M  75M S 17.2   5.8   0:01.68 io.appium.uiaut+
```
 
(_The above was gathered with:_ `while 1; do adb shell top -m 1000 -d 1 -n 1 | grep "<pid>"; sleep 1; done`)

---

### Pixel 4 (Real Device):

#### Screenshot capture `averageDurationMs`

```bash
04-15 17:51:11.417  7912  7937 I System.out: averageDurationMs over 30 screenshot(s): 93.36667
04-15 17:51:14.454  7912  7937 I System.out: averageDurationMs over 30 screenshot(s): 78.166664
04-15 17:51:17.508  7912  7937 I System.out: averageDurationMs over 30 screenshot(s): 85.13333
04-15 17:51:20.585  7912  7937 I System.out: averageDurationMs over 30 screenshot(s): 83.666664
04-15 17:51:23.584  7912  7937 I System.out: averageDurationMs over 30 screenshot(s): 78.166664
04-15 17:51:26.597  7912  7937 I System.out: averageDurationMs over 30 screenshot(s): 81.166664
04-15 17:51:29.610  7912  7937 I System.out: averageDurationMs over 30 screenshot(s): 79.53333
04-15 17:51:32.614  7912  7937 I System.out: averageDurationMs over 30 screenshot(s): 79.53333
04-15 17:51:35.630  7912  7937 I System.out: averageDurationMs over 30 screenshot(s): 80.166664
04-15 17:51:38.712  7912  7937 I System.out: averageDurationMs over 30 screenshot(s): 75.53333
```

(_The above was gathered by instrumenting:_ `matchFramerate` _in:_ MjpegScreenshotStream.java)

#### Current `master` - CPU & Memory

```bash
11208 u0_a244      20   0 4.8G  95M  58M S  0.0   1.7   0:03.96 io.appium.uiaut+
11208 u0_a244      20   0 4.8G  95M  58M S  3.7   1.7   0:03.97 io.appium.uiaut+
11208 u0_a244      20   0 4.8G  95M  58M S  3.7   1.7   0:04.01 io.appium.uiaut+
11208 u0_a244      20   0 4.8G  95M  58M S  0.0   1.7   0:04.03 io.appium.uiaut+
11208 u0_a244      20   0 4.8G  95M  58M S  0.0   1.7   0:04.05 io.appium.uiaut+
```

#### With PR #340 applied (0 client(s)) - CPU & Memory

```bash
11889 u0_a244      20   0 4.9G  95M  57M S  0.0   1.7   0:00.64 io.appium.uiaut+
11889 u0_a244      20   0 4.9G  95M  57M S  0.0   1.7   0:00.65 io.appium.uiaut+
11889 u0_a244      20   0 4.9G  95M  57M S  0.0   1.7   0:00.65 io.appium.uiaut+
11889 u0_a244      20   0 4.9G  95M  57M S  0.0   1.7   0:00.65 io.appium.uiaut+
11889 u0_a244      20   0 4.9G  95M  57M S  3.7   1.7   0:00.65 io.appium.uiaut+
```

#### With PR #340 applied (1 client(s)) - CPU & Memory

```bash
11889 u0_a244      20   0 4.9G 105M  60M S 51.8   1.9   0:01.17 io.appium.uiaut+
11889 u0_a244      20   0 4.9G 107M  60M S 44.4   1.9   0:01.81 io.appium.uiaut+
11889 u0_a244      20   0 4.9G 116M  67M S 48.1   2.1   0:02.44 io.appium.uiaut+
11889 u0_a244      20   0 4.9G 109M  66M S 48.1   1.9   0:03.08 io.appium.uiaut+
11889 u0_a244      20   0 4.9G 101M  61M S 48.1   1.8   0:03.68 io.appium.uiaut+
```

(_The above was gathered with:_ `while 1; do adb shell top -m 1000 -d 1 -n 1 | grep "<pid>"; sleep 1; done`)

mykola-mokhnach commented:
yep, we get about 83 ms average screenshot duration on a real device, which gives us maximum 12 FPS. And this is Pixel 4, which is one of the top-performing devices. I was getting minimum 150-180ms per frame on some older Samsung devices though. Also, for cases where on-screen animations were running or the app under test in general was not idling for longer time Android forces instrumentation to wait, which was producing no frames at all in the broadcasting stream. This means the resulting video would have noticeable "freezes" while one is doing, for example, swipes/scrolls or is watching videos. Have you tried such scenarios?

jzaleski commented:
Yes at times you can notice the lower frame-rate though I wouldn't call it freezes or lag exactly because it's intentionally capped to 10 FPS.

For a little more context, many of our use-cases involve testing CRUD apps, mostly on newer devices / newer API versions, using simulators. For manual test cases 10-20 FPS has proven to be more than sufficient for our target audience. For automation, FPS really doesn't matter that much.

Longer term as hardware / software continues to evolve I think this approach will probably become more feasible. In the near to mid-term it meets the need for a large portion of the market, but is clearly not viable for mobile games, streaming, video testing, etc.

Maybe we can treat this as an experimental feature targeted at bringing parity between the Android / iOS solutions for mjpeg out-of-the-box, caveated w/ a note to the effect of: "for higher intensity workflows it would be better to use these alternatives" (the list you provided above)?

---

(In unrelated news, I cannot reproduce the CI failure(s) that are currently occurring locally -- I'm trying to figure out what might be going on there [now])

mykola-mokhnach commented:
> Yes at times you can notice the lower frame-rate though I wouldn't call it freezes or lag exactly because it's intentionally capped to 10 FPS.

10 FPS is not an issue here. I'm fine with that. I'm more worried about non-idling instrumentation delays and related streaming drops. Could you please answer to my recent question - does this implementation experience the same issue I described above?

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/model/settings/MjpegFiltering.java:
```diff
@@ -0,0 +1,37 @@
+
+import io.appium.uiautomator2.server.ServerConfig;
+
+public class MjpegFiltering extends AbstractSetting<Boolean> {
```
Please add short docstring to each setting describing the purpose of it and acceptable value ranges

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/ServerConfig.java:
```diff
@@ -16,10 +16,106 @@
-    private final static int PORT = 6790;
+    // server port (default: `6790`)
+    private final static int SERVER_PORT =
+        System.getenv("SERVER_PORT") == null ? 6790 :
```
these all ternary operators could be moved to a helper method

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/ServerConfig.java:
```diff
@@ -16,10 +16,106 @@
 
     public static int getServerPort() {
-        return PORT;
+        if (overrides.containsKey("serverPort")) {
```
with concurrent map it does not work like that. Please use `getOrDefault` method instead.

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/ServerConfig.java:
```diff
@@ -16,10 +16,106 @@
     public static int getServerPort() {
-        return PORT;
+        if (overrides.containsKey("serverPort")) {
+            return (int) overrides.get("serverPort");
```
please avoid using magic strings. Use setting name property if possible

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/ServerInstrumentation.java:
```diff
@@ -73,20 +76,39 @@ private void setAccessibilityServiceState() {
         this.serverPort = serverPort;
+
+        if (!isValidPort(mjpegServerPort)) {
+            throw new UiAutomator2Exception(String.format(
```
I'd rather simply log a warning here if we want to make mjpeg server less intrusive

jzaleski replied:
Good call, this will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotClient.java:
```diff
@@ -0,0 +1,112 @@
+        try {
+            String line = in.readLine();
+            while (line == null) {
+                line = in.readLine();
```
this loop might hug CPU. Consider checking `available` property first

jzaleski replied:
The `available` method does not exist on `BufferedReader` however `ready` does and achieves the same goal, allowing me to add a `sleep` while in the busy loop.

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotClient.java:
```diff
@@ -0,0 +1,112 @@
+                "Screenshot broadcast starting for %s",
+                getRemoteAddress()
+            ));
+            String start = "HTTP/1.0 200 OK\r\nServer: AQI Android Screenshot Socket Server\r\nConnection: close\r\nMax-Age: 0\r\nExpires: 0\r\nCache-Control: no-cache, private\r\nPragma: no-cache\r\nContent-Type: multipart/x-mixed-replace; boundary=--BoundaryString\r\n\r\n";
```
Please move the magic string into constants

mykola-mokhnach replied:
or we could even move the bytes of it to a constant

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotClient.java:
```diff
@@ -0,0 +1,112 @@
+        try {
+            out.write(data);
+        } catch (IOException e) {
+            Logger.info("Client socket connection not writable. Closing... ", e);
```
warning

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotClient.java:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    void closeSocket() {
+        try {
```
we should probably short circuit here if `closed` is already `true`

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotClient.java:
```diff
@@ -0,0 +1,112 @@
+
+class MjpegScreenshotClient {
+    private final Socket socket;
+    private boolean closed = false;
```
I usually add `is` prefix to boolean properties

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotStream.java:
```diff
@@ -0,0 +1,142 @@
+            }
+
+            // how long each loop should take in milliseconds
+            long targetInterval =
```
consider switching to ScheduledThreadPoolExecutor -> scheduleAtFixedRate

jzaleski replied:
With the changed made to synchronization, short-circuiting / minimizing busy-loops, I think we're in pretty good shape.

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotServer.java:
```diff
@@ -0,0 +1,91 @@
+
+public class MjpegScreenshotServer extends Thread {
+    private final List<MjpegScreenshotClient> clients =
+        Collections.synchronizedList(new ArrayList<MjpegScreenshotClient>());
```
👍

jzaleski replied:
Removed this in favor of just synchronizing on the object where necessary

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotStream.java:
```diff
@@ -0,0 +1,142 @@
+    }
+
+    private byte[] getScreenshot() {
+        Bitmap screenshot = uia.takeScreenshot();
```
this method may return null

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotStream.java:
```diff
@@ -0,0 +1,142 @@
+
+        byte[] header = String.format(
+            Locale.ROOT,
+            "--BoundaryString\r\nContent-type: image/jpg\r\nContent-Length: %d\r\n\r\n",
```
Please move the magic string to constants

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotStream.java:
```diff
@@ -0,0 +1,142 @@
+            "--BoundaryString\r\nContent-type: image/jpg\r\nContent-Length: %d\r\n\r\n",
+            jpeg.length
+        ).getBytes(UTF_8);
+        byte[] end = "\r\n\r\n".getBytes(UTF_8);
```
⬆️

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/utils/ScreenshotHelper.java:
```diff
@@ -147,17 +158,51 @@ public static String takeScreenshot() throws TakeScreenshotException {
 
+    public static byte[] compressJpeg(final Bitmap bitmap, float scale, int quality, boolean filter) throws TakeScreenshotException {
+        Bitmap resultBitmap;
+        if (scale == 1.0) {
```
we cannot compare float number for equality. Check https://stackoverflow.com/questions/1088216/whats-wrong-with-using-to-compare-floats-in-java

jzaleski replied:
It seems like overkill, but I've taken the suggestion from the referenced link and this will be resolved w/ the upcoming update(s)

mykola-mokhnach replied:
image scaling hits performance. This means that the scaling will always be performed does not matter which scale factor we provide with the current implementation.

jzaleski replied:
That's correct, to your point, I'd like to avoid unnecessary CPU hits if at all possible.

jzaleski replied:
Understood.

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/utils/ScreenshotHelper.java:
```diff
@@ -147,17 +158,51 @@ public static String takeScreenshot() throws TakeScreenshotException {
         }
     }
 
+    public static byte[] compressJpeg(final Bitmap bitmap, float scale, int quality, boolean filter) throws TakeScreenshotException {
```
As far as I can read in Android documentation Jpeg compression is the most resource-intensive operation in general. What if we try to use the low level MediaCodec interfaces to produce Mjpeg stream: https://github.com/archersmind/EncoderDecoderTestForMJPEG/blob/master/EncoderDecoderTestForMJPEG.java

jzaleski replied:
I think this seems like a great future optimization.

jzaleski replied:
Marking as resolved.

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotStream.java:
```diff
@@ -0,0 +1,142 @@
+
+            byte[] screenshotData = getScreenshot();
+            if (screenshotData.length > 0) {
+                MjpegScreenshotClient client;
```
client could be defined inside the loop

jzaleski replied:
I don't think there is any harm in reassigned the reference. Unless you feel really strongly about this, I'd like to leave it as is.

mykola-mokhnach replied:
Let the compiler take care about it. Luckily these are smart enough nowadays to not care too much about minor optimisations. Although, from the readability and code structure perspective it is always better to keep all variable definitions inside blocks where these are used and as close as possible to the place they are used

jzaleski replied:
OK, no problem, and you're right. I guess I'm showing my age :-P

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotStream.java:
```diff
@@ -0,0 +1,142 @@
+            byte[] screenshotData = getScreenshot();
+            if (screenshotData.length > 0) {
+                MjpegScreenshotClient client;
+                for (Iterator<MjpegScreenshotClient> iterator = clients.iterator(); iterator.hasNext();) {
```
I think we could simply perform the loop inside `synchronized (clients)` block and make the list of clients itself a simple ArrayList, like it is done for iOS. This change will also require all add/remove operations on this list to be enclosed into such block as well.
See https://www.pixelstech.net/article/1449133298-Java-SynchronizedList-and-Iterator for more details

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotStream.java:
```diff
@@ -0,0 +1,142 @@
+
+        if (duration < targetInterval) {
+            try {
+                sleep(targetInterval - duration);
```
Consider using SystemClock.sleep instead

jzaleski replied:
It seems reasonable to use `SystemClock` in both instances, thanks for the suggestion.

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/ServerConfig.java:
```diff
@@ -16,10 +16,106 @@
+        System.getenv("MJPEG_SERVER_FRAMERATE") == null ? 10 :
+            Integer.parseInt(System.getenv("MJPEG_SERVER_FRAMERATE"));
+    // mjpeg scale should be between 1 and 100 (default: `50`)
+    private final static int MJPEG_SCALING_FACTOR =
```
I would rather consider the actual screen density to calculate the default scale factor. By default uia takes sceenshots having `display.getRealSize(displaySize)` dimensions.

jzaleski replied:
To me this is configuration available to the integrator. While I agree basing it off the screen-density, etc. could be feasible I'd like to keep it as an absolute value.

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/ServerConfig.java:
```diff
@@ -16,10 +16,106 @@
+            Integer.parseInt(System.getenv("MJPEG_SCALING_FACTOR"));
+    // mjpeg quality is between 0 and 100 (default: `75`)
+    private final static int MJPEG_SERVER_SCREENSHOT_QUALITY =
+        System.getenv("MJPEG_SERVER_SCREENSHOT_QUALITY") == null ? 75 :
```
75 is too much for a default value. 50 or even lower is enough

jzaleski replied:
OK. I'll follow your guidance on this one.

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/model/settings/MjpegFiltering.java:
```diff
@@ -0,0 +1,37 @@
+import io.appium.uiautomator2.server.ServerConfig;
+
+public class MjpegFiltering extends AbstractSetting<Boolean> {
+    private static final String SETTING_NAME = "mjpegFiltering";
```
how about renaming this to `mjpegUseBilinearFiltering`?

jzaleski replied:
That seems like a better / more descriptive name

jzaleski replied:
Going to to drop "use" from it however, because it looks really weird for the getters / setters

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/model/settings/MjpegScalingFactor.java:
```diff
@@ -0,0 +1,37 @@
+
+    @Override
+    protected void apply(Integer value) {
+        ServerConfig.setMjpegScalingFactor(value);
```
Do we want to apply a validation for setting values?

jzaleski replied:
It seems like most of the other `Setting(s)` **do not** validate. I think the doc-string comment is a reasonable starting point. Longer term it may make sense to add a `validate` method to the `AbstractSetting` class.

mykola-mokhnach replied:
I'm ok with that, but it usually makes sense to at least check if the given value is in the acceptable range and reset it to default/throw exception if not in case these values are received from the client (people sometimes don't know what they are doing and not reading the documentation ;) )

jzaleski replied:
Fair enough -- This will be resolved w/ the upcoming update(s)

jzaleski commented:
> yep, we get about 83 ms average screenshot duration on a real device, which gives us maximum 12 FPS. And this is Pixel 4, which is one of the top-performing devices. I was getting minimum 150-180ms per frame on some older Samsung devices though. Also, for cases where on-screen animations were running or the app under test in general was not idling for longer time Android forces instrumentation to wait, which was producing no frames at all in the broadcasting stream. This means the resulting video would have noticeable "freezes" while one is doing, for example, swipes/scrolls or is watching animations. Have you tried such scenarios?

If I'm understanding the ask correctly, no I have not noticed such delays in my testing.

jzaleski commented:
@mykola-mokhnach I have applied a large number of the suggestions, I appreciate the feedback / knowledge sharing.

As a side-effect of all the changes, it seems as though the CPU utilization is now a bit better as well (which was expected to some extent):

#### 0 client(s) - CPU & Memory

```bash
15264 u0_a90       20   0 1.3G  92M  59M S  0.0   6.1   0:00.97 io.appium.uiaut+
15264 u0_a90       20   0 1.3G  92M  59M S  0.0   6.1   0:00.98 io.appium.uiaut+
15264 u0_a90       20   0 1.3G  92M  59M S  0.0   6.1   0:00.99 io.appium.uiaut+
```

#### 1 client(s) - CPU & Memory

```bash
15264 u0_a90       20   0 1.3G  94M  60M S 19.2   6.3   0:01.04 io.appium.uiaut+
15264 u0_a90       20   0 1.3G  95M  60M S 19.2   6.3   0:01.30 io.appium.uiaut+
15264 u0_a90       20   0 1.3G  95M  60M S 23.0   6.3   0:01.58 io.appium.uiaut+
15264 u0_a90       20   0 1.3G  95M  60M S 19.2   6.3   0:01.87 io.appium.uiaut+
15264 u0_a90       20   0 1.3G  95M  60M S 19.2   6.3   0:02.15 io.appium.uiaut+
```

(**gathered with:** `while 1; do adb shell top -m 1000 -d 1 -n 1 | grep "appium"; sleep 1; done`)

Thanks for your review.

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/model/settings/MjpegBilinearFiltering.java:
```diff
@@ -0,0 +1,45 @@
+import io.appium.uiautomator2.server.ServerConfig;
+
+/**
+ * Controls whether or not bilinear filtering is enabled for streamed
```
I'd rather say: 

whether or not to apply bilinear filtering to source screenshot resize algorithm. Enabling this flag may improve the quality of the resulting scaled bitmap, but this introduces a small performance hit.

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/model/settings/MjpegScalingFactor.java:
```diff
@@ -0,0 +1,44 @@
+import io.appium.uiautomator2.server.ServerConfig;
+
+/**
+ * Controls the scaling factor of streamed screenshots
```
I'd rather say:

.. where the best quality is 100 and the worst is 1. The greater the value is the more CPU time is needed to encode a single bitmap into JPEG format. Usually the value between 25 and 90 is fine for this setting. Greater values affect performance too much without visible quality improvements and lower ones introduce visible distortions into the resulting image.

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/model/settings/MjpegServerFramerate.java:
```diff
@@ -0,0 +1,44 @@
+import io.appium.uiautomator2.server.ServerConfig;
+
+/**
+ * Controls the framerate of streamed screenshots
```
Greater framerate values create greater CPU load. That actual maximum framerate value is limited by the performance of the device under test

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/model/settings/MjpegServerScreenshotQuality.java:
```diff
@@ -0,0 +1,44 @@
+import io.appium.uiautomator2.server.ServerConfig;
+
+/**
+ * Controls the quality of streamed screenshots
```
I'd rather say:

.. where the best quality is 100 and the worst is 1. The greater the value is the more CPU time is needed to encode a single bitmap into JPEG format. Usually the value between 25 and 90 is fine for this setting. Greater values affect performance too much without visible quality improvements and lower ones introduce visible distortions into the resulting image.

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/model/settings/MjpegScalingFactor.java:
```diff
@@ -0,0 +1,44 @@
+import io.appium.uiautomator2.server.ServerConfig;
+
+/**
+ * Controls the scaling factor of streamed screenshots
```
I'd rather say:

100 means no scaling is applied (100% of the original size). The bigger image size is the more CPU performance is needed to encode it.

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/ServerConfig.java:
```diff
@@ -16,10 +16,131 @@
+
 public class ServerConfig {
-    private final static int PORT = 6790;
+    // server port (default: `6790`)
```
These comments are redundant

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/ServerConfig.java:
```diff
@@ -16,10 +16,131 @@
+
+    private static Integer getValueFromEnvOrDefault(String key, Integer defaultValue) {
+        String value = System.getenv(key);
+        if (value == null) {
```
a ternary operator usage could be more readable here and below

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/ServerConfig.java:
```diff
@@ -16,10 +16,131 @@
 
     public static int getServerPort() {
-        return PORT;
+        return (int) getValueFromOverridesOrDefault(
```
Please make the method call generic to avoid casting

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/ServerConfig.java:
```diff
@@ -16,10 +16,131 @@
+            MJPEG_BILINEAR_FILTERING);
+    }
+
+    /* Setter(s) */
```
I usually use regions notation to make it work better with IDEs. See https://blog.jetbrains.com/idea/2012/03/custom-code-folding-regions-in-intellij-idea-111/

jzaleski replied:
I just removed the comments since they really weren't adding a ton of value / clarification over the obvious.

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/ServerInstrumentation.java:
```diff
@@ -73,20 +76,39 @@ private void setAccessibilityServiceState() {
+                mjpegServerPort
+            ));
+        }
+        this.mjpegServerPort = mjpegServerPort;
```
should we still set it if it is not in valid range?

jzaleski replied:
I thought about this when I was making the logging vs. exception change from the previous comments and think yes, because since both the `{start,stop}MjpegServer` method(s)  are now guarded w/ the `isValidPort` call as well.

mykola-mokhnach replied:
I mean maybe it makes sense to just keep the default value if the retrieved one is invalid?

jzaleski replied:
This will be resolved w/ the upcoming update(s)

jzaleski replied:
Sure, let's go w/ that. I'll update the logging accordingly.

jzaleski replied:
I did the same for `serverPort` as well for consistency -- please let me know what you think once I update.

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotClient.java:
```diff
@@ -0,0 +1,123 @@
+
+class MjpegScreenshotClient {
+    private static final byte[] START = 
+        "HTTP/1.0 200 OK\r\nServer: AQI Android Screenshot Socket Server\r\nConnection: close\r\nMax-Age: 0\r\nExpires: 0\r\nCache-Control: no-cache, private\r\nPragma: no-cache\r\nContent-Type: multipart/x-mixed-replace; boundary=--BoundaryString\r\n\r\n".getBytes(UTF_8);
```
this string is too long. Pease split it into lines

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotClient.java:
```diff
@@ -0,0 +1,123 @@
+            this.in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
+            this.out = socket.getOutputStream();
+        } catch (IOException e) {
+            Logger.error("Client failed to initialize");
```
it is a good practice to add exceptions to log messages: `Logger.error("Client failed to initialize", e)`

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotClient.java:
```diff
@@ -0,0 +1,123 @@
+        }
+
+        try {
+            while (!in.ready()) {
```
👍

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotServer.java:
```diff
@@ -0,0 +1,99 @@
+    private final MjpegScreenshotStream mjpegScreenshotStream =
+        new MjpegScreenshotStream(clients);
+    private boolean isStopped = false;
+    private int port;
```
final?

jzaleski replied:
Updated references to things that will not be re-assigned across all the Mjpeg* classes.

This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotServer.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    private void addClient(MjpegScreenshotClient client) {
+        synchronized (clients) {
```
😻

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotStream.java:
```diff
@@ -0,0 +1,140 @@
+public class MjpegScreenshotStream extends Thread {
+    private static final String BOUNDARY_STRING = 
+        "--BoundaryString\r\nContent-type: image/jpg\r\nContent-Length: %d\r\n\r\n";
+    private static byte[] END = "\r\n\r\n".getBytes(UTF_8);
```
final

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotStream.java:
```diff
@@ -0,0 +1,140 @@
+    @Override
+    public void run() {
+        while (!isStopped) {
+            if (clients.size() == 0) {
```
please synchronize

also `clients.size() == 0` -> clients.isEmpty()

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotStream.java:
```diff
@@ -0,0 +1,140 @@
+    public void run() {
+        while (!isStopped) {
+            if (clients.size() == 0) {
+               SystemClock.sleep(500);
```
please format

mykola-mokhnach replied:
the interval could also be moved to constants

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotStream.java:
```diff
@@ -0,0 +1,140 @@
+            if (screenshotData.length > 0) {
+                synchronized (clients) {
+                    for (MjpegScreenshotClient client : clients) {
+                        if (client.isClosed()) {
```
we should eventually remove closed clients from the list

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotStream.java:
```diff
@@ -0,0 +1,140 @@
+            long start = System.currentTimeMillis();
+
+            byte[] screenshotData = getScreenshot();
+            if (screenshotData.length > 0) {
```
this will fail if screenshotData is null

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotServer.java:
```diff
@@ -0,0 +1,99 @@
+            return;
+        }
+
+        mjpegScreenshotStream.start();
```
I'm not sure how exceptions in the mjpegScreenshotStream are handled. Let say what happens if jpeg compression throws an exception there?

jzaleski replied:
I've wrapped the `getScreenshot` call, the one w/ the most potential to raise, w/ a `try` / `catch` and we'll just drop that frame.

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/model/settings/MjpegBilinearFiltering.java:
```diff
@@ -40,6 +42,9 @@ public Boolean getValue() {
     @Override
     protected void apply(Boolean value) {
+        if (value == null) {
+            throw new UiAutomator2Exception(String.format("Invalid %s value specified", SETTING_NAME));
```
There is a dedicated exception for Illegal arguments

mykola-mokhnach replied:
same below

jzaleski replied:
OK. I'll update them all.

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/model/settings/MjpegScalingFactor.java:
```diff
@@ -39,6 +42,9 @@ public Integer getValue() {
     @Override
     protected void apply(Integer value) {
+        if (value == null || value < 1 || value > 100) {
+            throw new UiAutomator2Exception(String.format("Invalid %s value specified", SETTING_NAME));
```
Please also include the description of the expected value ranges into the error message. Something like:

```
throw new IllegalArgumentException(String.format("The scaling factor value must be in range 1..100. %s was given instead", value));
```

mykola-mokhnach replied:
same below

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/ServerConfig.java:
```diff
@@ -27,44 +27,45 @@
     }
 
-    private static Object getValueFromOverridesOrDefault(String key, Object defaultValue) {
+    private static <T> T getValueFromOverridesOrDefault(String key, T defaultValue) {
```
👍

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/ServerInstrumentation.java:
```diff
@@ -78,24 +78,30 @@ private void setAccessibilityServiceState() {
         if (!isValidPort(serverPort)) {
-            throw new UiAutomator2Exception(String.format(
-                "The server port is out of valid range [%s;%s]: %s",
+            Logger.warn(String.format(
```
warn is deprecated. Please use `warning`

jzaleski replied:
`warning` doesn't seem to be defined.

mykola-mokhnach replied:
ah, nvm, just ignore them

jzaleski replied:
Logger is generally referring to: app/src/main/java/io/appium/uiautomator2/utils/Logger.java which currently only has a `warn`.

mykola-mokhnach replied:
yep, I have forgotten that this project uses a custom logging wrapper

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/ServerInstrumentation.java:
```diff
@@ -78,24 +78,30 @@ private void setAccessibilityServiceState() {
 
     private ServerInstrumentation(Context context, int serverPort, int mjpegServerPort) {
         if (!isValidPort(serverPort)) {
```
positive conditions are easier to read than negative. Consider flipping it

mykola-mokhnach replied:
same below

jzaleski replied:
Sure.

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotClient.java:
```diff
@@ -30,14 +30,23 @@
-    private static final byte[] START = 
-        "HTTP/1.0 200 OK\r\nServer: AQI Android Screenshot Socket Server\r\nConnection: close\r\nMax-Age: 0\r\nExpires: 0\r\nCache-Control: no-cache, private\r\nPragma: no-cache\r\nContent-Type: multipart/x-mixed-replace; boundary=--BoundaryString\r\n\r\n".getBytes(UTF_8);
+    private static final int INPUT_NOT_READY_SLEEP_TIME_MS = 100;
+    private static final byte[] START = (
```
👍

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotClient.java:
```diff
@@ -30,14 +30,23 @@
+    private static final int INPUT_NOT_READY_SLEEP_TIME_MS = 100;
+    private static final byte[] START = (
+        "HTTP/1.0 200 OK\r\n" +
+        "Server: AQI Android Screenshot Socket Server\r\n" +
```
what is AQI?

jzaleski replied:
Looks like that's a hold-over from our org name Applause App Quality, Inc. (AKA "AQI") -- I'll remove.

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotClient.java:
```diff
@@ -87,14 +96,14 @@ void initialize() {
         try {
             while (!in.ready()) {
-               SystemClock.sleep(100);
+               SystemClock.sleep(INPUT_NOT_READY_SLEEP_TIME_MS);
```
👍

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotStream.java:
```diff
@@ -33,8 +33,11 @@
     private static final String BOUNDARY_STRING = 
-        "--BoundaryString\r\nContent-type: image/jpg\r\nContent-Length: %d\r\n\r\n";
-    private static byte[] END = "\r\n\r\n".getBytes(UTF_8);
+        "--BoundaryString\r\n" +
```
shouldn't the boundary string be some random value?

jzaleski replied:
The string / constant portion you've referenced is not random, but once it's merged w/ the length of the `jpeg` (below). I think what we have is fine, the spec is here: https://tools.ietf.org/html/rfc7578#section-4.1

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotStream.java:
```diff
@@ -53,8 +56,8 @@ public void interrupt() {
         while (!isStopped) {
-            if (clients.size() == 0) {
-               SystemClock.sleep(500);
+            if (clients.isEmpty()) {
```
please add synchronisation. Something like:

```
boolean hasNoClients;
synchronized (clients) {
  hasNoClients = clients.isEmpty();
}
if (hasNoClients) ...
```

jzaleski replied:
This will be resolved w/ the upcoming update(s)

jzaleski commented:
Just finished addressing the latest round of code-review comments.

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/model/settings/MjpegBilinearFiltering.java:
```diff
@@ -0,0 +1,54 @@
+    protected void apply(Boolean value) {
+        if (value == null) {
+            throw new InvalidArgumentException(String.format(
+                "Invalid %s value specified, must be false|true. %s was given",
```
nit: here `%s` is always `null`

jzaleski replied:
Thanks for keeping me honest :-P

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotStream.java:
```diff
@@ -0,0 +1,160 @@
+            try {
+                screenshotData = getScreenshot();
+            } catch (Exception e) {
+                screenshotData = new byte[0];
```
should we put the actual exception into the log?

jzaleski replied:
Yeah I think so, it would help w/ troubleshooting

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotStream.java:
```diff
@@ -0,0 +1,160 @@
+                        client.write(screenshotData);
+                    }
+                }
+                matchFramerate(targetInterval, start);
```
should we still wait if fails to get a screenshot?

jzaleski replied:
Yeah I think that makes sense. Updating.

mykola-mokhnach commented:
How do you think - does it make sense to have at least one integration test for screenshot scaling/compression helpers?

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotStream.java:
```diff
@@ -0,0 +1,165 @@
+                        client.write(screenshotData);
+                    }
+                }
+            }
```
here in else block it also makes sense to leave some logs

jzaleski replied:
I don't think the `else` block is entirely necessary here, because most of the failure modes will come from exceptions being raised / caught and there is already logging there. But just to cover our based I've added a message.

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotStream.java:
```diff
@@ -0,0 +1,165 @@
+    private byte[] getScreenshot() {
+        Bitmap screenshot = UI_AUTOMATION.takeScreenshot();
+        if (screenshot == null) {
+            return new byte[0];
```
how about throwing an exception instead, so we know why there where no screenshot?

jzaleski replied:
Sure, that's fine.

jzaleski replied:
This will be resolved w/ the upcoming update(s)

mykola-mokhnach commented on app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotStream.java:
```diff
@@ -0,0 +1,168 @@
+    private byte[] getScreenshot() {
+        Bitmap screenshot = UI_AUTOMATION.takeScreenshot();
+        if (screenshot == null) {
+            throw new TakeScreenshotException("Could not take screenshot");
```
we could also write `Could not take screenshot: UIA returned none`. It is possible that the current application view has `secure` flag set, so this line would constantly be logged while such view is visible.

jzaleski replied:
This will be resolved w/ the upcoming update(s)

jzaleski commented:
I just rebased current `master` -- and applied the latest round of suggestions.

---

As an aside, for the last few builds I've noticed that the `"DeviceCommandsTest"` has had one failure, somewhere in here: https://github.com/appium/appium-uiautomator2-server/blob/master/app/src/androidTestE2eTest/java/io/appium/uiautomator2/unittest/test/DeviceCommandsTest.java#L77

I've tried, but I **cannot** reproduce the issue locally, [however] it seems to fail pretty consistently on _Travis_ at the same place. I also don't know how I could have possibly broken it w/ my change-set.

Thoughts?

jzaleski commented:
Actually I tried again and I was able to reproduce the failure(s) -- the 2 commented out blocks below are the failures on API 26:

```bash
diff --git a/app/src/androidTestE2eTest/java/io/appium/uiautomator2/unittest/test/DeviceCommandsTest.java b/app/src/androidTestE2eTest/java/io/appium/uiautomator2/unittest/test/DeviceCommandsTest.java
index a3e2975..9369137 100644
--- a/app/src/androidTestE2eTest/java/io/appium/uiautomator2/unittest/test/DeviceCommandsTest.java
+++ b/app/src/androidTestE2eTest/java/io/appium/uiautomator2/unittest/test/DeviceCommandsTest.java
@@ -87,20 +87,20 @@ public class DeviceCommandsTest extends BaseTest {
         response = findElement(by);
         assertFalse(by + " should not be found", response.isSuccessful());
 
-        by = By.xpath("//*[@resource-id='android:id/action_bar']");
-        response = findElement(by);
-        assertTrue(response.isSuccessful());
+        // by = By.xpath("//*[@resource-id='android:id/action_bar']");
+        // response = findElement(by);
+        // assertTrue(response.isSuccessful());
 
         by = By.xpath("(//*[@class='android.widget.TextView'])[3]");
         response = findElement(by);
         response = getAttribute(response.getElementId(), "text");
         assertEquals("Accessibility", response.getValue());
 
-        by = By.xpath("//*[@resource-id='android:id/content']" +
-                "//*[@resource-id='android:id/text1'][5]");
-        response = findElement(by);
-        response = getAttribute(response.getElementId(), "text");
-        assertEquals("Content", response.getValue());
+        // by = By.xpath("//*[@resource-id='android:id/content']" +
+        //         "//*[@resource-id='android:id/text1'][5]");
+        // response = findElement(by);
+        // response = getAttribute(response.getElementId(), "text");
+        // assertEquals("Content", response.getValue());
```

jzaleski commented:
In my latest commit, I commented out the two tests I found to be failing -- I'd prefer to fix them in this change-set, but I think I might need a bit of help tracking down the root-cause.

jzaleski commented:
My hunch was correct, by prefixing `findElementTest` w/ `"zzz"`, which forced the test to run first, I was able to get consistent result(s).

As of now, this seems like a good enough work-around.

KazuCocoa commented on app/src/androidTestE2eTest/java/io/appium/uiautomator2/unittest/test/DeviceCommandsTest.java:
```diff
@@ -69,12 +69,11 @@
      */
     @Test
-    public void findElementTest() {
+    public void zzzFindElementTest() {
```
Could you leave a comment on why `zzz` is here?
https://github.com/appium/appium-uiautomator2-server/pull/340#issuecomment-616229653

jzaleski replied:
Absolutely. Hopping on a call in a few minutes, but I have some free time immediately after that which should be enough to get both the comment here and the markdown updated.

Thanks for your review(s)

jlipps replied:
probably a comment and a `// TODO figure out test state ordering issue`

jzaleski replied:
This will be resolved w/ the upcoming update(s)

jzaleski commented:
> thanks for this contribution @jzaleski! have you tested whether this mjpeg stream works with appium's `mjpegScreenshotUrl` capability? i.e., is the type of mjpeg stream produced by this server compatible with the mjpeg client appium uses? it would be ideal if so. i tried to add mjpeg server support based on @mykola-mokhnach's gstreamer changes, and it "worked", but the mjpeg frames sent by the server were not readable by appium's mjpeg client (and i haven't had a chance to dig into why). so all that to say, i'm curious if this just works.

@jlipps will test this after I complete the docs post-merge of this PR. Thanks for the suggestion, I agree it would be great if it "just worked" or required minimal tweaking 😃

jzaleski commented:
@mykola-mokhnach  / @KazuCocoa added a comment to `DeviceCommandsTest#findElementTest` -- as requested, I will update https://github.com/appium/appium/blob/master/docs/en/advanced-concepts/settings.md and test @jlipps suggestion post-merge.

jzaleski commented:
Created a PR for the documentation changes here: https://github.com/appium/appium/pull/14203 -- thankfully it seems like the CI job is finally running here 😃

jzaleski commented:
Alrighty, CI here and for the related documentation update to _Appium_ has passed at this point.

KazuCocoa commented:
appium-uiautomator2-server@4.8.0
